### PR TITLE
Fixed bug where dev-provided certificate was not being attached to client assertion

### LINKED ISSUE #7082: acquireTokenByClientCredential broken for clientCertificate
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

2.8.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

n/a

### Public or Confidential Client?

Confidential

### Description

I upgraded to the latest version (2.8 from 2.7) and my call to acquireTokenByClientCredential fails now.

### Error Message

{
    "errorCode": "invalid_client",
    "errorMessage": "7000216 - [2024-05-07 15:55:13Z]: AADSTS7000216: 'client_assertion', 'client_secret' or 'request' is required for the 'client_credentials' grant type. Trace ID: d1a0baab-1cf3-4a20-a263-605f65ceb200 Correlation ID: a93c308b-63ff-4880-8d07-383a27828034 Timestamp: 2024-05-07 15:55:13Z - Correlation ID: a93c308b-63ff-4880-8d07-383a27828034 - Trace ID: d1a0baab-1cf3-4a20-a263-605f65ceb200",
    "subError": "",
    "name": "ServerError",
    "errorNo": 7000216,
    "correlationId": "2bedce44-4ef7-4752-abb9-39cfe913cea3"
}

### MSAL Logs

_No response_

### Network Trace (Preferrably Fiddler)

- [ ] Sent
- [ ] Pending

### MSAL Configuration

```javascript
const clientConfig: Configuration = {
      auth: {
                clientCertificate: {
                    thumbprint,
                    privateKey
                },
                clientId: ,
                authority
            }
        }
```

### Relevant Code Snippets

```javascript
const pca = new ConfidentialClientApplication(clientConfig);
        const clientCredentialRequest: ClientCredentialRequest = {
            scopes: ["https://graph.microsoft.com/.default"]
        };
        const newToken = pca
            .acquireTokenByClientCredential(clientCredentialRequest)
```

### Reproduction Steps

Run the code above

### Expected Behavior

A token is returned

### Identity Provider

Entra ID (formerly Azure AD) / MSA

### Browsers Affected (Select all that apply)

Chrome

### Regression

@azure/msal-node @2.7.0

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**
Fixes [acquireTokenByClientCredential broken for clientCertificate #7082](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/7082)

Applied boy-scout-rule to `ConfidentialClientApplication.spec.ts` (contains unit tests). I've been waiting for a good opportunity to do this. The ConfidentialClientApplication tests are now in line with the other test files: All Managed Identity sources, ClientCredentialClient, OnBehalfOfClient and UsernamePasswordClient.

### CODE REVIEW

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 1]
```diff
@@ -229,206 +194,247 @@ describe("ConfidentialClientApplication", () => {
...
+            const clientAssertion: ClientAssertion = await getClientAssertionSpy
+                .mock.results[0].value;
+            expect(clientAssertion.assertion.length).toBeGreaterThan(1);
+            expect(clientAssertion.assertionType).toBe(
+                Constants.JWT_BEARER_ASSERTION_TYPE
+            );
+        });
     });
 
-    test("acquireTokenByClientCredential handles AuthErrors as expected", async () => {
```
Comment (Robbie-Microsoft) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts:389]: Deleted the test `acquireTokenByClientCredential handles AuthErrors as expected`.

This test doesn't accomplish anything

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 2]
```diff
@@ -229,206 +194,247 @@ describe("ConfidentialClientApplication", () => {
...
-        ).rejects.toMatchObject(
-            createClientAuthError(ClientAuthErrorCodes.missingTenantIdError)
-        );
+            const clientAssertion: ClientAssertion = await getClientAssertionSpy
+                .mock.results[0].value;
+            expect(clientAssertion.assertion.length).toBeGreaterThan(1);
+            expect(clientAssertion.assertionType).toBe(
+                Constants.JWT_BEARER_ASSERTION_TYPE
+            );
+        });
```
Comment (Robbie-Microsoft) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts]: This is the relevant unit test I added for the bug fix

File: lib/msal-node/test/client/ClientTestUtils.ts [hunk 3]
```diff
@@ -357,6 +359,53 @@ export class ClientTestUtils {
...
+            loggerCallback: (): void => {},
+            piiLoggingEnabled: true,
+            logLevel: LogLevel.Verbose,
+        };
+
+        const confidentialClientConfig: Configuration = {
+            auth: {
+                clientId: TEST_CONSTANTS.CLIENT_ID,
+                authority: TEST_CONSTANTS.AUTHORITY,
+                clientSecret: TEST_CONSTANTS.CLIENT_SECRET,
```
Comment (bgavrilMS) [inline lib/msal-node/test/client/ClientTestUtils.ts]: nit: client secret should not be the default. cert is more secure.

File: lib/msal-node/test/utils/CryptoKeys.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import crypto from "crypto";
+
+export class CryptoKeys {
+    private _thumbprint: string;
+    public get thumbprint(): string {
+        return this._thumbprint;
+    }
+    private set thumbprint(value: string) {
```
Comment (bgavrilMS) [inline lib/msal-node/test/utils/CryptoKeys.ts]: nit: you don't need private setters. Let the ctor assign directly to the field.

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**